### ISSUE TITLE
Worker thread dies with exception in 2.0.1

### ISSUE THREAD

**Issue by OpakAlex** [elastic/apm-agent-ruby#237]
[ElasticAPM] Queue is full (256 items), skipping…
What this means? Why it happens? 

0 new records into index since update version.

---

Update by @mikker: See [this comment](https://github.com/elastic/apm-agent-ruby/issues/237#issuecomment-481738046) for possible settings to tweak. I'm working on a docs page describing possible remedies to full queues.

**Comment by mikker** [comment]
Hi @OpakAlex!

It means the agent has trouble sending its recorded transactions to APM Server.

Your logs will

**Comment by OpakAlex** [comment]
@mikker trying to find something, only in production mode.
Also i have some warnings:
```
[8] ! WARNING: Detected 2 Thread(s) started in app boot:
[8] ! #&lt;Thread:0x000055787bfba218@/bundle/ruby/2.5.0/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:297 sleep&gt; - /bundle/ruby/2.5.0/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:299:in `sleep'
[8] ! #&lt;Thread:0x000055787c139cd8@/bundle/ruby/2.5.0/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:317 sleep_forever&gt; - /bundle/ruby/2.5.0/gems/elastic-apm-2.0.1/lib/elastic_apm/transport/worker.rb:44:in `pop'
```

**Comment by OpakAlex** [comment]
@mikker do i need on apm server turn on `v2`?

**Comment by mikker** [comment]
Yes, version 2+ of the agent requires APM Server to be at least version 6.5.

Looking into the warnings.

**Comment by OpakAlex** [comment]
@mikker i want to understand this.

**Comment by mikker** [comment]
Yes, APM Server 6.5 should work with ES all the way back to 6.2, see https://www.elastic.co/support/matrix#matrix_compatibility under _Product Compatibility_

**Comment by roncohen** [comment]
@OpakAlex please be aware that even though APM Server 6.5 can work with Elasticsearch 6.2, you will need to use Kibana 6.5 to visualize trace data from APM Server 6.5

**Comment by OpakAlex** [comment]
So, APM can write correct to ES 6.2, but if i need to see this results correct, i need kibana 6.5, right?
For kibana 6.5 i need ES 6.5 right? 
@roncohen @mikker

**Comment by mikker** [comment]
👍 That's correct.

**Comment by mikker** [comment]
That's an error in the docs actually – default level is `INFO`. Sorry about that.

**Comment by Zatvobor** [comment]
Guys, it's crap. You are there for enjoying, its clear for me now. May I ask you to put the notice about your fun into project README.

**Comment by mikker** [comment]
Hi @Zatvobor. Do you mean a notice about compatibility between versions?

**Comment by Zatvobor** [comment]
Never mind, we decided to remove APM from our hosting schema.

**Comment by mikker** [comment]
@Zatvobor I'll be happy to help you get it working. Is your issue the same as @OpakAlex's?

**Comment by OpakAlex** [comment]
@mikker deploing log level debug, in few minutes i will have some logs. It's looks like issue with  puma webserver. let me see some logs.

**Comment by mikker** [comment]
It could be an issue with Puma booting the agent's worker threads before starting its own threads.

I have a branch that lazy boots the worker threads – maybe try that on your next deploy? https://github.com/elastic/apm-agent-ruby/tree/lazy-boot-workers

**Comment by Zatvobor** [comment]
> @Zatvobor I'll be happy to help you get it working. Is your issue the same as @OpakAlex's?

May I ask you to release `v1.1.1` with fixed issue #224 (#198)

**Comment by OpakAlex** [comment]
```
[ElasticAPM] [2.0.1] Starting agent, reporting to http://apm-server.production.svc.cluster.local:8200
[ElasticAPM] Booting 1 workers
[ElasticAPM] Starting instrumenter
```
[ElasticAPM] Queue is full (256 items), skipping…
No more logs.

**Comment by OpakAlex** [comment]
@mikker i use this `preload_app!` in my puma config. 
So looks like issue here, do you think your `lazy-boot-workers` branch can fix it?
```
# Use the `preload_app!` method when specifying a `workers` number.
# This directive tells Puma to first boot the application and load code
# before forking the application. This takes advantage of Copy On Write
# process behavior so workers use less memory. If you use this option
# you need to make sure to reconnect any threads in the `on_worker_boot`
# block.
#
```

**Comment by mikker** [comment]
I think it might 👍

**Comment by OpakAlex** [comment]
looks better @mikker  :
```
[ElasticAPM] Bytes sent during this request: 18232
[ElasticAPM] Bytes sent during this request: 18907
[ElasticAPM] Bytes sent during this request: 19206
[ElasticAPM] Closing request
[ElasticAPM] APM Server responded with status 202
```

**Comment by OpakAlex** [comment]
but @mikker where is view for requests? http://take.ms/FN0lf

**Comment by mikker** [comment]
@Zatvobor The issue you mention will be fixed and backported to the v1.x series as well.

**Comment by Zatvobor** [comment]
awesome. Thanks

**Comment by OpakAlex** [comment]
@mikker hello, it's me again ;)
So we fixed for puma, but it broken for sidekiq ;)

**Comment by OpakAlex** [comment]
can you make env option for this method? Then i can put into docker image

**Comment by mikker** [comment]
How is it broken for Sidekiq? Did it stop reporting with the lazy workers?

Do you mean something like `ELASTIC_APM_EAGER_BOOT_WORKERS`? I'd rather fix it everywhere so it just works 😉

**Comment by OpakAlex** [comment]
like this: https://github.com/elastic/apm-agent-ruby/pull/243

**Comment by OpakAlex** [comment]
yes same errors for sidekiq containers: 
`[ElasticAPM] Queue is full (256 items), skipping…`

**Comment by mikker** [comment]
Does your PR fix both your problems?

**Comment by OpakAlex** [comment]
will see in 5 min

**Comment by mikker** [comment]
Beer is fair – owe you one 😄

The actual sending of the data happens in the workers, off the main thread. So we are not blocking requests actually. Sending events from the main app via Sidekiq (or other job runners) is not something we are working on.

**Comment by OpakAlex** [comment]
ok. not a problem. will be at Denmark i will ping you ;)

**Comment by OpakAlex** [comment]
@mikker looks like issue still happens.
```
 WARN: [ElasticAPM] Queue is full (256 items), skipping…
```

It working for 1 hour, then nothing into APM.
Any ideas? It's sidekiq part...

**Comment by OpakAlex** [comment]
```
      def work_forever
        while (msg = queue.pop)
          case msg
          when StopMessage
            stop
          else
            process msg
          end

          next unless stopping?

          debug 'Stopping worker -- %s', self
          @connection.flush
          break
        end
      rescue Exception => e
        warn 'Worker died with exception: %s', e.inspect
        debug e.backtrace.join("\n")
      end
```
can be possible, if we have 0 in queue, this loop finished? `queue.pop` => nil ?

**Comment by mikker** [comment]
`Queue#pop` blocks the thread until it finds something in the queue, so not that.

But it looks like the workers die after the one hour mark and then have trouble getting going again.

I run both Puma and Sidekiq in an app myself and have no trouble. There's no error messages in your Sidekiq logs? Perhaps do the `DEBUG`level there if you haven't already?

**Comment by OpakAlex** [comment]
will do debug level. and then look into issues. so so strange. but it happens after 1hour +-. we have some load.

**Comment by OpakAlex** [comment]
`WARN: [ElasticAPM] Queue is full (256 items), skipping…`

Nothing more on debugg level @mikker 
Ideas?

**Comment by OpakAlex** [comment]
after 57 minutes...

**Comment by OpakAlex** [comment]
`But it looks like the workers die after the one hour` @mikker how they can die?

**Comment by mikker** [comment]
They die if they come across an exception, which they then log. Then on the next submitted event they are (or at least should be) restarted.

**Comment by mikker** [comment]
Here's the logging code:
https://github.com/elastic/apm-agent-ruby/blob/master/lib/elastic_apm/transport/worker.rb#L58-L61

**Comment by OpakAlex** [comment]
http://take.ms/YXh6z

**Comment by OpakAlex** [comment]
@mikker can we run method itself here: https://github.com/elastic/apm-agent-ruby/blob/master/lib/elastic_apm/transport/worker.rb#L60, if exception try to run `work_forever `, like run loop?

**Comment by OpakAlex** [comment]
@mikker ping, i am here again ;)
So my fix with run loop works fine. But it's a tmp solution. What you think about?

**Comment by mikker** [comment]
Hmm, I think this sounds like the problem is rooted in something like the Puma case. I'll investigate tomorrow.

**Comment by OpakAlex** [comment]
thanks!

**Comment by OpakAlex** [comment]
just to let you known, my fix works good, i have metrics for 8 hours. @mikker

**Comment by mikker** [comment]
Sounds good!

I still think it sounds like your problem is actually something else. Your GZip streams are getting closed too soon and we don't know why. I'm holding off on merging your fix while I investigate a bit more. Thanks again for helping out!

**Comment by OpakAlex** [comment]
thanks man!

**Comment by dpankros** [comment]
FYI, I just saw this error too.  Also using puma.

**Comment by mikker** [comment]
@OpakAlex Saw you closed your PR. Did you try the new `2.1.1` release to see if it fixes your problems?

**Comment by OpakAlex** [comment]
i tried 2.1.1, but the main problem with sidekiq exists. I need to understand how you restart thred.

**Comment by OpakAlex** [comment]
and why when i put loop run, i have memory leaks. This is main issue, if you can explain this, i will try to find a fix.

**Comment by mikker** [comment]
The threads ought to restart on the next event, see https://github.com/elastic/apm-agent-ruby/blob/master/lib/elastic_apm/transport/base.rb#L37.

The leak could be the main thread reaching this https://github.com/elastic/apm-agent-ruby/blob/master/lib/elastic_apm/transport/base.rb#L70 and starting another worker on the next event, while your workers keep restarting themselves. Essentially booting a new, soon-to-be rogue worker on every event?

**Comment by OpakAlex** [comment]
@mikker why here https://github.com/elastic/apm-agent-ruby/blob/master/lib/elastic_apm/transport/base.rb#L39 we just return nil? I have this message on puma too, but it works... it means we lost something into apm?

**Comment by mikker** [comment]
Yes it means the APM agent/server can't keep up with your app's load. You could try increasing the agent's worker pool size and see if that helps?

**Comment by OpakAlex** [comment]
yes, trying thanks.

**Comment by alvarolobato** [comment]
Hi @OpakAlex do you have any update on this issue? Are you still facing the same situation?

**Comment by OpakAlex** [comment]
0 updates, it works some hours, then stopped. I don't have ideas.

**Comment by phlegx** [comment]
We are having same issues. 

```[ElasticAPM] Worker died with exception: #<Zlib::GzipFile::Error: closed gzip stream>```

But seems to happen only in Sidekiq workers. The application tracing itself works fine.

**Comment by tak1n** [comment]
Hey there I'm working for the same company as @phlegx 

We figured out it also happens for a normal rails server process (not only sidekiq processes).

Funnily enough it didn't appear yet for a rails 4 app we have (puma server nor sidekiq) only for our rails 5 app.

**Comment by mikker** [comment]
Hi @tak1n! Bummed to hear this is still an issue for you. Are you on the latest version of the agent?

**Comment by OpakAlex** [comment]
@mikker i updated all system yesterday.... ruby 2.6.1, rails 5.2.2. Also from master gem....

**Comment by tak1n** [comment]
Hi @mikker, yea we are using elastic-apm 2.3.1

**Comment by tak1n** [comment]
Short update form our side:

We started to use it in production and there we didn't see such exceptions, only thing which happens from time to time is:

`[ElasticAPM] Queue is full (256 items), skipping…`

We tried to increase the pool size like suggested in the documentation (currently using 20).
The interesting thing here is that it works for quite some time and then randomly we get spammed by this. 

In APM server logs all requests are handled properly and it responds with 202.

After restarting the production apps in question everything works again.

So it seems like some kind of deadlock on the ruby side, but not sure.

**Comment by tak1n** [comment]
So after restarting now we see alot of:

```
[ElasticAPM] Worker died with exception: #<Zlib::GzipFile::Error: closed gzip stream>
```

Sorry to say, but we have to currently stop using elastic_apm for that matter.
If there is anything I can help with let me know.

**Comment by mikker** [comment]
Thank you for updating the thread when you know more 👍
That error could also mean the worker has died and thereby stopped emptying the queue. Ie the same problem. One of my todos is to make sure we get better logs for situations like this.

It's unfortunate that the agent keeps dying for you. I'll let you know if I make any new discoveries.

**Comment by OpakAlex** [comment]
Same for me: http://take.ms/ZydpI
After last update workers die...
Any ideas @mikker why?

**Comment by mikker** [comment]
No ideas. But we need a better way to find the actual exceptions happening in your setups.

**Comment by OpakAlex** [comment]
Last setup it's just update ruby and rails. @mikker how is work about better logs? Can we add more logs? Then i can try it in my setup. Because now i have 7h / 24h only. And it's looks like pain.

**Comment by danilvoe** [comment]
@mikker  One gets the impression that the agent cannot resume the transfer of data to the server after the server was not available. Is it possible?

**Comment by mikker** [comment]
@danilvoe No, that should not be the case. It opens up a new connection on every try.

**Comment by phlegx** [comment]
@mikker  We are sending the APM stuff trough our internal VPN (OpenVPN) maybe this is causing these issues. Do you perhaps know if that might be an issue?

**Comment by monsha** [comment]
Same issues as @phlegx 

We had lot of `[ElasticAPM] Worker died with exception: #<Zlib::GzipFile::Error: closed gzip stream>` then Sidekiq processes were killed, after update the gem to 2.6.1 the Zlib error went away, but still happens `WARN: [ElasticAPM] Queue is full (256 items), skipping…` there is a lot of them

Ruby 2.5.1
Rails 5.2.0
apm-server 6.5.4

**Comment by mikker** [comment]
Hi @monsha. Do you enqueue a bunch of jobs at the same time?

There are three things you can try:
- Bump `config.api_buffer_size` to something greater than 256 and see if that helps.
- Bump `config.pool_size` to spawn more worker threads to empty the queue faster.
- Set `config.transaction_sample_rate` to something less than `1.0`. This will send fewer spans but works great if you have high load.

Good to hear the Gzip error went away. I'm finding that one especially hard to debug.

**Comment by phlegx** [comment]
@mikker I tried to increase the Queue size `config.api_buffer_size` to `1500`, however my Rails app (I did not restart it) still shows:

`[ElasticAPM] Queue is full (256 items), skipping…`

Any ideas why? 

How can I see how much items are in the queue?

thanks in advance.

**Comment by OpakAlex** [comment]
Hi guys, looks like i known answer! Testing into our production, and will write tomorrow to you ;))

**Comment by mikker** [comment]
What a cliffhanger! 😊

**Comment by OpakAlex** [comment]
so, my monitoring shows like this for 12 hours:
http://take.ms/Scd9s

looks like it works!

**Comment by mikker** [comment]
Great! What did you find out?

**Comment by OpakAlex** [comment]
so first one my rails config:
```
pool_size: 10
api_buffer_size: 512
```
Then i changed APM server config:
`output.elasticsearch.workers: 5`
`output.elasticsearch.bulk_max_size: 5120 `
`queue.mem.events: 5 * 5120`

I have 3 apm-server instances, and one cluster with 8 es nodes.

**Comment by OpakAlex** [comment]
i will continue monitoring today. we will see.

**Comment by mikker** [comment]
Nice to know, @OpakAlex. Is it still going strong?

**Comment by OpakAlex** [comment]
works as "children going to school ;)"
http://take.ms/unvac

**Comment by monsha** [comment]
@mikker after the changes (commented by @OpakAlex above)
`[ElasticAPM] Queue is full (256 items), skipping…` went away but now it's not a warn is an error: `ERROR: [ElasticAPM] APM Server responded with an error: "queue is full"`, why now it's an errors instead of warn?, and how to find/avoid this error?

**Comment by danilvoe** [comment]
> so first one my rails config:
> 
> ```
> pool_size: 10
> api_buffer_size: 512
> ```
> 
> Then i changed APM server config:
> `output.elasticsearch.workers: 5`
> `output.elasticsearch.bulk_max_size: 5120 `
> `queue.mem.events: 5 * 5120`
> 
> I have 3 apm-server instances, and one cluster with 8 es nodes.

can you share the whole apm server config? Can't repeat your success :(

**Comment by OpakAlex** [comment]
@danilvoe my apm config as soon as default, just this only points i changed.
But which error do you have?

**Comment by OpakAlex** [comment]
also @mikker can you look:
```
error handling request	{"request_id": "f4f4f803-c7f0-44e9-9b1f-2f751b6e2106", "method": "POST", "URL": "/intake/v2/events", "content_length": -1, "remote_address": "10.244.12.174", "user-agent": "http.rb/4.0.3", "error": "event exceeded the permitted size. [{\"span\":{\"id\":\"6449b958bfe0f109\",\"transaction_id\":\"c71fdf369e3794fe\",\"parent_id\":\"c71fdf369e3794fe\",\"name\":\"SELECT FROM mentions\",\"type\":\"db.postgresql.sql\",\"duration\":11701.315,\"context\":{\"sync\":true,\"db\":{\"instance\":null,\"statement\":\"SELECT \\\"mentions\\\".\\\"id\\\" FROM \\\"mentions\\\" WHERE \\\"mentions\\\".\\\"organization_id\\\" = 273 AND \\\"mentions\\\".\\\"deleted_at\\\" IS NULL AND \\\"mentions\\\".\\\"response_id\\\" IN (38651924, 38651851, 26622600, 26622773, 29340136, 26615763, 26623402, 26623421, 26624058, 30042751, 26624807, 26623390, 26624386, 26625996, 25599426, 44914385, 26634058, 25592700, 32081331, 26638297, 26638341, 26638339, 25588382, 26641540, 26641581, 26641599, 25588288, 26641568, 25633219, 25596232, 30050562, 30050583, 30050588, 25600478, 25606071, 25607111, 25607123, 25618700, 30050704, 25633573, 30050714, 25637319, 25670575, 25685466, 30050808, 29138123, 26641628, 25600773, 25602052, 31364681, 26642033, 26642269, 30051585, 26642521, 26642522, 26642692, 26654545, 26650618, 26650634, 26642850, 26642897, 26652706, 31278289, 26642862, 32081343, 32081565, 26638340, 27155524, 27156202, 27155408, 26643219, 26643233, 26643249, 26645008, 26643467, 28313921, 28331585, 28324981, 28324976, 28315306, 26643552, 28315349, 26653001, 27024958, 28312563, 28258645, 28267263, 28259367, 28346667, 28256221, 26644069, 27127338, 27122084, 27117355, 27104887, 27108428, 26653553, 26644785, 26644874, 26644892, 28255180, 2825
```

**Comment by monsha** [comment]
I assume that APM server error, is related with internal queue in server so it should be fixed increasing `queue.mem.events`? let's see

**Comment by monsha** [comment]
@mikker there is a way to see how many items are in the queue or some metrics about it?
Thanks for you help

**Comment by danilvoe** [comment]
> @danilvoe my apm config as soon as default, just this only points i changed.
> But which error do you have?

`WARN -- : [ElasticAPM] Queue is full (256 items), skipping…`

**Comment by OpakAlex** [comment]
Sometimes this errors comes back... i ma debugging this guys. So strange.

**Comment by phlegx** [comment]
As I have written I increased the `config.api_buffer_size`  from  `256` to  to `1500`. But on my application, even when restarting it, at some point I still get the `256` queue full error.

```[ElasticAPM] Queue is full (256 items), skipping…```

For me as well would be really helpful to see the queue size and how many items are really queued, as @monsha was already asking.

**Comment by mikker** [comment]
The setting is used here https://github.com/elastic/apm-agent-ruby/blob/master/lib/elastic_apm/transport/base.rb#L17.

On holiday this week.

**Comment by mikker** [comment]
`master` has been updated with more robust (hopefully) connection handling – can you try running the agent from there and let me know if that fixes your issues?

**Comment by danilvoe** [comment]
After updating, the error has changed to

`E, [2019-05-20T11:34:48.658107 #1102] ERROR -- : [ElasticAPM] [THREAD:47537840]: APM Server not responding in time, terminating request`

On the side of the server apm problems are not visible.

**Comment by mikker** [comment]
@danilvoe What does your load look like? Your APM Server might be too busy. Perhaps dial the load a bit down with `transaction_sample_rate`?

**Comment by mikker** [comment]
Closing this issue as it has become a mess of all sorts of things. Feel free to open a new issue if you don't feel your issue was fixed.
